File: Repo/opencog/atomspace/c++/opencog-atomspace.cpp
/*
 * opencog-atomspace.cpp
 * 
 * OpenCog AtomSpace - Hypergraph Knowledge Representation
 * A single-file implementation of a hypergraph database for storing
 * and manipulating knowledge in a graph structure with typed nodes and links.
 * 
 * The AtomSpace is the core knowledge representation system in OpenCog,
 * allowing flexible representation of semantic networks, logical relationships,
 * and complex knowledge structures.
 */

#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <memory>
#include <set>
#include <sstream>
#include <algorithm>

namespace opencog {
namespace atomspace {

// Forward declarations
class Atom;
class Node;
class Link;
class AtomSpace;

using AtomPtr = std::shared_ptr<Atom>;
using NodePtr = std::shared_ptr<Node>;
using LinkPtr = std::shared_ptr<Link>;

// Atom types enumeration
enum AtomType {
    ATOM,
    NODE,
    LINK,
    CONCEPT_NODE,
    PREDICATE_NODE,
    VARIABLE_NODE,
    EVALUATION_LINK,
    INHERITANCE_LINK,
    SIMILARITY_LINK,
    LIST_LINK,
    AND_LINK,
    OR_LINK,
    NOT_LINK
};

// Base Atom class - represents any entity in the hypergraph
class Atom {
protected:
    AtomType type;
    float truthValue;  // Simple truth value [0.0, 1.0]
    float confidence;  // Confidence in truth value [0.0, 1.0]

public:
    Atom(AtomType t) : type(t), truthValue(1.0f), confidence(1.0f) {}
    virtual ~Atom() {}

    AtomType getType() const { return type; }
    float getTruthValue() const { return truthValue; }
    float getConfidence() const { return confidence; }

    void setTruthValue(float tv, float conf = 1.0f) {
        truthValue = tv;
        confidence = conf;
    }

    virtual std::string toString() const = 0;
    virtual bool isNode() const { return false; }
    virtual bool isLink() const { return false; }

    static std::string typeToString(AtomType type) {
        switch(type) {
            case ATOM: return "Atom";
            case NODE: return "Node";
            case LINK: return "Link";
            case CONCEPT_NODE: return "ConceptNode";
            case PREDICATE_NODE: return "PredicateNode";
            case VARIABLE_NODE: return "VariableNode";
            case EVALUATION_LINK: return "EvaluationLink";
            case INHERITANCE_LINK: return "InheritanceLink";
            case SIMILARITY_LINK: return "SimilarityLink";
            case LIST_LINK: return "ListLink";
            case AND_LINK: return "AndLink";
            case OR_LINK: return "OrLink";
            case NOT_LINK: return "NotLink";
            default: return "Unknown";
        }
    }
};

// Node class - represents named entities
class Node : public Atom {
private:
    std::string name;

public:
    Node(AtomType type, const std::string& nodeName) 
        : Atom(type), name(nodeName) {}

    const std::string& getName() const { return name; }
    bool isNode() const override { return true; }

    std::string toString() const override {
        std::stringstream ss;
        ss << "(" << typeToString(type) << " \"" << name << "\" ";
        ss << "tv=" << truthValue << " conf=" << confidence << ")";
        return ss.str();
    }
};

// Link class - represents relationships between atoms
class Link : public Atom {
private:
    std::vector<AtomPtr> outgoing;

public:
    Link(AtomType type, const std::vector<AtomPtr>& atoms)
        : Atom(type), outgoing(atoms) {}

    const std::vector<AtomPtr>& getOutgoing() const { return outgoing; }
    size_t getArity() const { return outgoing.size(); }
    bool isLink() const override { return true; }

    std::string toString() const override {
        std::stringstream ss;
        ss << "(" << typeToString(type);
        for (const auto& atom : outgoing) {
            ss << "\n  " << atom->toString();
        }
        ss << ")";
        return ss.str();
    }
};

// AtomSpace class - the hypergraph container
class AtomSpace {
private:
    std::map<std::string, NodePtr> nodeIndex;  // Index nodes by name+type
    std::vector<AtomPtr> atoms;  // All atoms in the space
    std::map<AtomPtr, std::set<LinkPtr>> incomingIndex;  // Incoming links per atom

    std::string makeNodeKey(AtomType type, const std::string& name) const {
        return Atom::typeToString(type) + ":" + name;
    }

public:
    AtomSpace() {}

    // Add a node to the AtomSpace
    NodePtr addNode(AtomType type, const std::string& name) {
        std::string key = makeNodeKey(type, name);
        
        // Return existing node if already present
        auto it = nodeIndex.find(key);
        if (it != nodeIndex.end()) {
            return it->second;
        }

        // Create new node
        NodePtr node = std::make_shared<Node>(type, name);
        nodeIndex[key] = node;
        atoms.push_back(node);
        return node;
    }

    // Add a link to the AtomSpace
    LinkPtr addLink(AtomType type, const std::vector<AtomPtr>& outgoing) {
        LinkPtr link = std::make_shared<Link>(type, outgoing);
        atoms.push_back(link);

        // Update incoming index
        for (const auto& atom : outgoing) {
            incomingIndex[atom].insert(link);
        }

        return link;
    }

    // Get incoming links for an atom
    std::set<LinkPtr> getIncoming(const AtomPtr& atom) const {
        auto it = incomingIndex.find(atom);
        return (it != incomingIndex.end()) ? it->second : std::set<LinkPtr>();
    }

    // Query atoms by type
    std::vector<AtomPtr> getAtomsByType(AtomType type) const {
        std::vector<AtomPtr> result;
        for (const auto& atom : atoms) {
            if (atom->getType() == type) {
                result.push_back(atom);
            }
        }
        return result;
    }

    // Get node by name and type
    NodePtr getNode(AtomType type, const std::string& name) const {
        std::string key = makeNodeKey(type, name);
        auto it = nodeIndex.find(key);
        return (it != nodeIndex.end()) ? it->second : nullptr;
    }

    // Get all atoms
    const std::vector<AtomPtr>& getAllAtoms() const {
        return atoms;
    }

    // Get count of atoms
    size_t getSize() const {
        return atoms.size();
    }

    // Print all atoms
    void print() const {
        std::cout << "AtomSpace contains " << atoms.size() << " atoms:" << std::endl;
        for (const auto& atom : atoms) {
            std::cout << "  " << atom->toString() << std::endl;
        }
    }

    // Clear the AtomSpace
    void clear() {
        atoms.clear();
        nodeIndex.clear();
        incomingIndex.clear();
    }
};

} // namespace atomspace
} // namespace opencog

// Demonstration of AtomSpace functionality
int main() {
    using namespace opencog::atomspace;

    std::cout << "OpenCog AtomSpace - Hypergraph Knowledge Representation Demo" << std::endl;
    std::cout << "============================================================" << std::endl << std::endl;

    // Create an AtomSpace
    AtomSpace atomspace;

    std::cout << "Creating a simple knowledge base..." << std::endl << std::endl;

    // Create concept nodes
    auto human = atomspace.addNode(CONCEPT_NODE, "human");
    auto mortal = atomspace.addNode(CONCEPT_NODE, "mortal");
    auto socrates = atomspace.addNode(CONCEPT_NODE, "Socrates");
    auto animal = atomspace.addNode(CONCEPT_NODE, "animal");

    std::cout << "Created nodes:" << std::endl;
    std::cout << "  " << human->toString() << std::endl;
    std::cout << "  " << mortal->toString() << std::endl;
    std::cout << "  " << socrates->toString() << std::endl;
    std::cout << "  " << animal->toString() << std::endl << std::endl;

    // Create inheritance relationships
    auto link1 = atomspace.addLink(INHERITANCE_LINK, {socrates, human});
    auto link2 = atomspace.addLink(INHERITANCE_LINK, {human, mortal});
    auto link3 = atomspace.addLink(INHERITANCE_LINK, {human, animal});

    std::cout << "Created inheritance links:" << std::endl;
    std::cout << "  " << link1->toString() << std::endl;
    std::cout << "  " << link2->toString() << std::endl;
    std::cout << "  " << link3->toString() << std::endl << std::endl;

    // Add truth values
    link1->setTruthValue(1.0f, 1.0f);  // Socrates is definitely human
    link2->setTruthValue(1.0f, 1.0f);  // All humans are mortal
    link3->setTruthValue(1.0f, 0.9f);  // Humans are animals (high confidence)

    // Create predicates
    auto breathes = atomspace.addNode(PREDICATE_NODE, "breathes");
    auto thinks = atomspace.addNode(PREDICATE_NODE, "thinks");

    // Create evaluation links
    auto eval1 = atomspace.addLink(EVALUATION_LINK, {breathes, socrates});
    auto eval2 = atomspace.addLink(EVALUATION_LINK, {thinks, socrates});

    eval1->setTruthValue(1.0f, 1.0f);
    eval2->setTruthValue(1.0f, 1.0f);

    std::cout << "Created evaluation links:" << std::endl;
    std::cout << "  " << eval1->toString() << std::endl;
    std::cout << "  " << eval2->toString() << std::endl << std::endl;

    // Query the knowledge base
    std::cout << "Querying the AtomSpace..." << std::endl << std::endl;

    std::cout << "All concept nodes:" << std::endl;
    auto concepts = atomspace.getAtomsByType(CONCEPT_NODE);
    for (const auto& atom : concepts) {
        std::cout << "  " << atom->toString() << std::endl;
    }
    std::cout << std::endl;

    std::cout << "All inheritance links:" << std::endl;
    auto inheritances = atomspace.getAtomsByType(INHERITANCE_LINK);
    for (const auto& atom : inheritances) {
        std::cout << "  " << atom->toString() << std::endl;
    }
    std::cout << std::endl;

    // Query incoming links
    std::cout << "Links involving 'human':" << std::endl;
    auto incoming = atomspace.getIncoming(human);
    for (const auto& link : incoming) {
        std::cout << "  " << link->toString() << std::endl;
    }
    std::cout << std::endl;

    // Display complete knowledge base
    std::cout << "Complete AtomSpace contents:" << std::endl;
    atomspace.print();
    std::cout << std::endl;

    // Demonstrate logical operations
    std::cout << "Creating logical expressions..." << std::endl;
    auto prop1 = atomspace.addNode(PREDICATE_NODE, "is_alive");
    auto prop2 = atomspace.addNode(PREDICATE_NODE, "is_wise");
    
    auto andLink = atomspace.addLink(AND_LINK, {prop1, prop2});
    std::cout << "  AND expression: " << andLink->toString() << std::endl;

    auto orLink = atomspace.addLink(OR_LINK, {prop1, prop2});
    std::cout << "  OR expression: " << orLink->toString() << std::endl;

    std::cout << std::endl;
    std::cout << "Final AtomSpace size: " << atomspace.getSize() << " atoms" << std::endl;
    std::cout << "AtomSpace demonstration complete!" << std::endl;

    return 0;
}

File: Repo/opencog/atomspace/Lua/opencog-atomspace.lua
#!/usr/bin/env lua
--[[
opencog-atomspace.lua

OpenCog AtomSpace - Hypergraph Knowledge Representation in Lua

This single-file implementation demonstrates Lua's strengths for AI:
- Tables as flexible data structures for graphs
- Metatables for operator overloading (__len, __tostring)
- First-class functions for pattern matching
- Lightweight object system via prototypes
- Simple yet powerful semantics
]]

-- ===== Atom Types =====
-- Demonstrates: Tables as enums

local AtomType = {
    ATOM = "Atom",
    NODE = "Node",
    LINK = "Link",
    CONCEPT_NODE = "ConceptNode",
    PREDICATE_NODE = "PredicateNode",
    VARIABLE_NODE = "VariableNode",
    EVALUATION_LINK = "EvaluationLink",
    INHERITANCE_LINK = "InheritanceLink",
    SIMILARITY_LINK = "SimilarityLink",
    LIST_LINK = "ListLink",
    AND_LINK = "AndLink",
    OR_LINK = "OrLink",
    NOT_LINK = "NotLink"
}

-- ===== Truth Value =====
-- Demonstrates: Simple data objects

local TruthValue = {}
TruthValue.__index = TruthValue

function TruthValue:new(strength, confidence)
    local self = setmetatable({}, TruthValue)
    self.strength = strength or 1.0
    self.confidence = confidence or 1.0
    return self
end

function TruthValue:__tostring()
    return string.format("tv=%.2f conf=%.2f", self.strength, self.confidence)
end

-- ===== Atom (Base Class) =====
-- Demonstrates: Prototype-based inheritance

local Atom = {}
Atom.__index = Atom

local atom_id_counter = 0

function Atom:new(atom_type, tv)
    local self = setmetatable({}, Atom)
    atom_id_counter = atom_id_counter + 1
    self.id = atom_id_counter
    self.atom_type = atom_type
    self.tv = tv or TruthValue:new()
    return self
end

function Atom:get_type()
    return self.atom_type
end

function Atom:get_tv()
    return self.tv
end

function Atom:set_tv(tv)
    self.tv = tv
end

function Atom:__tostring()
    return string.format("Atom(%d, %s, %s)", self.id, self.atom_type, tostring(self.tv))
end

-- ===== Node =====
-- Demonstrates: Inheritance via metatables

local Node = setmetatable({}, {__index = Atom})
Node.__index = Node

function Node:new(atom_type, name, tv)
    local self = Atom:new(atom_type, tv)
    setmetatable(self, Node)
    self.name = name
    return self
end

function Node:get_name()
    return self.name
end

function Node:__tostring()
    return string.format("%s(\"%s\", %s)", self.atom_type, self.name, tostring(self.tv))
end

-- ===== Link =====
-- Demonstrates: Tables for collections

local Link = setmetatable({}, {__index = Atom})
Link.__index = Link

function Link:new(atom_type, outgoing, tv)
    local self = Atom:new(atom_type, tv)
    setmetatable(self, Link)
    self.outgoing = outgoing or {}
    return self
end

function Link:get_outgoing()
    return self.outgoing
end

function Link:get_arity()
    return #self.outgoing
end

function Link:__tostring()
    local names = {}
    for _, atom in ipairs(self.outgoing) do
        table.insert(names, tostring(atom))
    end
    return string.format("%s([%s], %s)", self.atom_type, table.concat(names, ", "), tostring(self.tv))
end

-- ===== AtomSpace =====
-- Demonstrates: Tables as hash maps, operator overloading

local AtomSpace = {}
AtomSpace.__index = AtomSpace

function AtomSpace:new()
    local self = setmetatable({}, AtomSpace)
    self.atoms = {}  -- atom_id -> atom
    self.nodes = {}  -- (type, name) -> atom
    self.links = {}  -- atom -> list of atoms
    return self
end

function AtomSpace:add_node(atom_type, name, tv)
    local key = atom_type .. ":" .. name
    
    if self.nodes[key] then
        return self.nodes[key]
    end
    
    local node = Node:new(atom_type, name, tv)
    self.atoms[node.id] = node
    self.nodes[key] = node
    
    return node
end

function AtomSpace:add_link(atom_type, outgoing, tv)
    local link = Link:new(atom_type, outgoing, tv)
    self.atoms[link.id] = link
    
    -- Build incoming sets
    for _, atom in ipairs(outgoing) do
        if not self.links[atom.id] then
            self.links[atom.id] = {}
        end
        table.insert(self.links[atom.id], link)
    end
    
    return link
end

function AtomSpace:get_atom(atom_id)
    return self.atoms[atom_id]
end

function AtomSpace:contains(atom)
    return self.atoms[atom.id] ~= nil
end

function AtomSpace:remove(atom)
    self.atoms[atom.id] = nil
    
    if atom.get_name then
        local key = atom.atom_type .. ":" .. atom:get_name()
        self.nodes[key] = nil
    end
    
    if atom.get_outgoing then
        for _, out_atom in ipairs(atom:get_outgoing()) do
            if self.links[out_atom.id] then
                local filtered = {}
                for _, link in ipairs(self.links[out_atom.id]) do
                    if link.id ~= atom.id then
                        table.insert(filtered, link)
                    end
                end
                self.links[out_atom.id] = filtered
            end
        end
    end
end

function AtomSpace:get_incoming(atom)
    return self.links[atom.id] or {}
end

function AtomSpace:clear()
    self.atoms = {}
    self.nodes = {}
    self.links = {}
    atom_id_counter = 0
end

function AtomSpace:size()
    local count = 0
    for _ in pairs(self.atoms) do
        count = count + 1
    end
    return count
end

function AtomSpace:get_all_atoms()
    local result = {}
    for _, atom in pairs(self.atoms) do
        table.insert(result, atom)
    end
    return result
end

function AtomSpace:dump()
    print(string.format("AtomSpace (size=%d):", self:size()))
    for _, atom in pairs(self.atoms) do
        print("  " .. tostring(atom))
    end
end

-- Operator overloading
function AtomSpace:__len()
    return self:size()
end

function AtomSpace:__tostring()
    return string.format("AtomSpace(size=%d)", self:size())
end

-- ===== Pattern Matching =====
-- Demonstrates: First-class functions, closures

function pattern_match(atomspace, pattern, callback)
    local matches = {}
    
    for _, atom in pairs(atomspace.atoms) do
        if pattern(atom) then
            table.insert(matches, atom)
            if callback then
                callback(atom)
            end
        end
    end
    
    return matches
end

-- ===== Query Functions =====

local function find_by_type(atomspace, atom_type)
    return pattern_match(atomspace, function(atom)
        return atom.atom_type == atom_type
    end)
end

local function find_by_name(atomspace, name)
    return pattern_match(atomspace, function(atom)
        return atom.get_name and atom:get_name() == name
    end)
end

local function find_links_with(atomspace, target_atom)
    return pattern_match(atomspace, function(atom)
        if atom.get_outgoing then
            for _, out_atom in ipairs(atom:get_outgoing()) do
                if out_atom.id == target_atom.id then
                    return true
                end
            end
        end
        return false
    end)
end

-- ===== Inference =====
-- Demonstrates: Graph traversal

local function transitive_closure(atomspace, start_atom)
    local visited = {}
    local result = {}
    
    local function visit(atom)
        if visited[atom.id] then
            return
        end
        
        visited[atom.id] = true
        table.insert(result, atom)
        
        local incoming = atomspace:get_incoming(atom)
        for _, link in ipairs(incoming) do
            if link.atom_type == AtomType.INHERITANCE_LINK then
                local outgoing = link:get_outgoing()
                if outgoing[1].id == atom.id and #outgoing >= 2 then
                    visit(outgoing[2])
                end
            end
        end
    end
    
    visit(start_atom)
    return result
end

-- ===== Demonstration Function =====

local function demonstrate_atomspace()
    print("=" .. string.rep("=", 69))
    print("OpenCog AtomSpace - Lua Implementation")
    print("=" .. string.rep("=", 69))
    print()
    
    -- Create atomspace
    print("1. Creating AtomSpace")
    print(string.rep("-", 50))
    local atomspace = AtomSpace:new()
    print(string.format("Created: %s", tostring(atomspace)))
    print()
    
    -- Add nodes
    print("2. Adding Nodes")
    print(string.rep("-", 50))
    local human = atomspace:add_node(AtomType.CONCEPT_NODE, "human")
    local mortal = atomspace:add_node(AtomType.CONCEPT_NODE, "mortal")
    local socrates = atomspace:add_node(AtomType.CONCEPT_NODE, "Socrates")
    local philosopher = atomspace:add_node(AtomType.CONCEPT_NODE, "philosopher")
    
    print(tostring(human))
    print(tostring(mortal))
    print(tostring(socrates))
    print(tostring(philosopher))
    print()
    
    -- Add links
    print("3. Adding Links (Relationships)")
    print(string.rep("-", 50))
    local link1 = atomspace:add_link(AtomType.INHERITANCE_LINK, {human, mortal})
    local link2 = atomspace:add_link(AtomType.INHERITANCE_LINK, {socrates, human})
    local link3 = atomspace:add_link(AtomType.INHERITANCE_LINK, {socrates, philosopher})
    
    print(tostring(link1))
    print(tostring(link2))
    print(tostring(link3))
    print()
    
    -- Size and containment
    print("4. AtomSpace Operations")
    print(string.rep("-", 50))
    print(string.format("AtomSpace size: %d atoms", atomspace:size()))
    print(string.format("Contains Socrates: %s", tostring(atomspace:contains(socrates))))
    print()
    
    -- Pattern matching
    print("5. Pattern Matching")
    print(string.rep("-", 50))
    local concept_nodes = find_by_type(atomspace, AtomType.CONCEPT_NODE)
    print(string.format("Found %d concept nodes:", #concept_nodes))
    for _, node in ipairs(concept_nodes) do
        print("  " .. tostring(node))
    end
    print()
    
    local inheritance_links = find_by_type(atomspace, AtomType.INHERITANCE_LINK)
    print(string.format("Found %d inheritance links:", #inheritance_links))
    for _, link in ipairs(inheritance_links) do
        print("  " .. tostring(link))
    end
    print()
    
    -- Incoming sets
    print("6. Incoming Sets (Graph Navigation)")
    print(string.rep("-", 50))
    local socrates_incoming = atomspace:get_incoming(socrates)
    print(string.format("Links pointing to Socrates (%d):", #socrates_incoming))
    for _, link in ipairs(socrates_incoming) do
        print("  " .. tostring(link))
    end
    print()
    
    -- Transitive closure
    print("7. Transitive Inference")
    print(string.rep("-", 50))
    print("Transitive closure starting from Socrates:")
    local closure = transitive_closure(atomspace, socrates)
    for _, atom in ipairs(closure) do
        print("  " .. tostring(atom))
    end
    print("→ Therefore, Socrates is mortal (via transitive inference)")
    print()
    
    -- Truth values
    print("8. Truth Values (Probabilistic Logic)")
    print(string.rep("-", 50))
    local uncertain = atomspace:add_node(AtomType.CONCEPT_NODE, "uncertain",
        TruthValue:new(0.7, 0.8))
    print(string.format("Node with uncertainty: %s", tostring(uncertain)))
    
    local weak_link = atomspace:add_link(AtomType.SIMILARITY_LINK,
        {human, philosopher}, TruthValue:new(0.5, 0.6))
    print(string.format("Link with weak similarity: %s", tostring(weak_link)))
    print()
    
    -- Dump atomspace
    print("9. Complete AtomSpace Dump")
    print(string.rep("-", 50))
    atomspace:dump()
    print()
    
    print("AtomSpace demonstration complete!")
    print("=" .. string.rep("=", 70))
    print("Lua strengths demonstrated:")
    print("  ✓ Tables as flexible graph structure")
    print("  ✓ Metatables for OOP and operator overloading")
    print("  ✓ First-class functions for pattern matching")
    print("  ✓ Closures for graph algorithms")
    print("  ✓ Lightweight and efficient implementation")
    print("  ✓ Prototype-based inheritance")
    print("=" .. string.rep("=", 70))
end

-- Run demonstration if executed directly
if arg and arg[0]:match("opencog%-atomspace%.lua$") then
    demonstrate_atomspace()
end

-- Export module
return {
    AtomType = AtomType,
    TruthValue = TruthValue,
    Atom = Atom,
    Node = Node,
    Link = Link,
    AtomSpace = AtomSpace,
    pattern_match = pattern_match,
    find_by_type = find_by_type,
    find_by_name = find_by_name,
    find_links_with = find_links_with,
    transitive_closure = transitive_closure
}
